crates/wasi/src/ctx.rs

```diff
@@ -417,7 +418,10 @@ impl WasiCtxBuilder {
     /// while returning `false` will reject the connection.
     pub fn socket_addr_check<F>(&mut self, check: F) -> &mut Self
     where
-        F: Fn(&SocketAddr, SocketAddrUse) -> bool + Send + Sync + 'static,
+        F: Fn(SocketAddr, SocketAddrUse) -> Pin<Box<dyn Future<Output = bool> + Send + Sync>>
```

Comment:
Since we're breaking this signature anyway, I also `&SocketAddr` -> `SocketAddr` (which is `Copy`).